pennylane/qnode_new/decorator.py

```diff
@@ -0,0 +1,126 @@
+# Copyright 2019 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+QNode decorator.
+"""
+from functools import lru_cache
+
+from .cv import CVQNode
+from .jacobian import JacobianQNode
+from .qnode import QNode as BaseQNode

```

Comment:
I believe that the ``QNode`` name for the function could be really confusing as it follows a class-like naming convention. Even so, currently this does stand for a class. Therefore looking exclusively at the following code (which is currently part of the decorator in the PR), one could think, that the ``__init__`` method of the ``QNode`` class is called:
```python
        return QNode(
            func,
            device,
            interface=interface,
            mutable=mutable,
            differentiable=differentiable,
            properties=properties,
        )
```

A name that refers in its first part to the creation of a qnode such as ``create_qnode`` could be better I believe (the word "construct" might be less preferable as per the terminology from other languages such as C++, though still fine I think).

Reply:
@antalszava, I agree, it's not ideal. The issue being that `QNode` _used_ to be a class, but now it is a function returning a class --- I gave it the same name for backwards compatibility (a lot of existing tutorials and documentation make use of it). I propose the following:

1. Rename the `QNode` class to `BaseQNode` (since it's no longer developer facing), as @co9olguy suggests.

2. Keep this function named as `QNode` for now, for backwards compatibility

3. Decide (in an ADR) if we want a new name for this function. If we do, include both names, but have the old one raise a deprecation warning, and remove it in a later release.

Reply:
I see, thanks! Backwards compatibility does seem like a good reason indeed. One thing that is not clear to me now, which examples would it break. What would be a simple example a break?

Would it actually not be enough to just rename this function have the if-else checks here and then return the specific QNode type (such that QNode stays the base class and each other QNode would be subclasses). This way the decorator would indeed return a QNode instance and the QNode signature does not have to majorly change either (as it could take a ``func`` and a ``device`` as before). Instead, this function would take further arguments to decide which QNode to return.

All in all, I agree that maybe this decision would need more attention and could be explored in an ADR potentially.

Reply:
More so that a lot of tests, tutorials, and documentation uses the `QNode()` class to construct qnodes currently. The aim is to make the qnode_new as compatible with the old class as possible to make the replacement easier.

Once the old QNode has been replaced, we can think about renaming (and it will be easier to make the change everywhere, all at once)